applications/MappingApplication/custom_searching/interface_object.h

```diff
@@ -0,0 +1,249 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Philipp Bucher, Jordi Cotela
+//
+// See Master-Thesis P.Bucher
+// "Development and Implementation of a Parallel
+//  Framework for Non-Matching Grid Mapping"
+
+#if !defined(KRATOS_INTERFACE_SEARCH_OBJECT_INCLUDED_H_INCLUDED )
+#define  KRATOS_INTERFACE_SEARCH_OBJECT_INCLUDED_H_INCLUDED
+
+
+// System includes
+
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+#include "includes/node.h"
+#include "geometries/geometry.h"
+
+
+namespace Kratos
+{
+///@addtogroup MappingApplication
+///@{
+
+///@name Kratos Classes
+///@{
+
+/// Object used by the bin-search
+/** This object is used by the bin search. It is the baseclass for objects that hold information
+ * about geometric entities, e.g. Nodes or Geometries
+*/
+class InterfaceObject : public Point
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of InterfaceObject
+    KRATOS_CLASS_POINTER_DEFINITION(InterfaceObject);
+
+    typedef Point BaseType;
+
+    typedef typename BaseType::CoordinatesArrayType CoordinatesArrayType;
+
+    typedef Node<3> NodeType;
+    typedef Kratos::shared_ptr<NodeType> NodePointerType;
+
+    typedef Geometry<NodeType> GeometryType;
+    typedef Kratos::shared_ptr<GeometryType> GeometryPointerType;
+
+    ///@}
+    ///@name  Enum's
+    ///@{
+
+    enum ConstructionType
+    {
+        Node_Coords,
+        Geometry_Center,
+        Element_Center,
+        Condition_Center
+    };
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    InterfaceObject(const CoordinatesArrayType& rCoordinates) : Point(rCoordinates)
+    { }
+
+    /// Destructor.
+    virtual ~InterfaceObject(){}
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void UpdateCoordinates(const CoordinatesArrayType& rCoordinates)
+    {
+        noalias(Coordinates()) = rCoordinates;
+    }
+
+    virtual void UpdateCoordinates()
+    {
+        KRATOS_ERROR << "Base class function called!" << std::endl;
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    virtual NodePointerType pGetBaseNode() const
```

Comment:
I don't fully understand the question
According to your suggestion the baseclass should return `nullptr`, but how would I actually return sth?

Reply:
The previous comment about returning a raw ptr instead of shared is if you plan/need to keep the function. My point here is that to use these functions, we need to find out what the derived type is similar to a dynamic cast:
```
if (pn = o.pGetBaseNode())
{
    do_something_with_node_object(pn);
}
else if (pg = o.pGetBaseGeometry())
{
   do_something_with_geom_object(pg);
}
```
We should avoid such base class methods alltogether if possible.

Reply:
Ah ok
these functions / derived objects are needed in order to make the search (bins-dynamics-objs) compatible with nodes or geometries (or elements)
I agree that this is not a good solution, but we tried already two years ago to find a better solution, but we failed

Anyway I will leave this as is, since this will be removed completely once the mpi-search anf the global-pointers are available

The case you are mentioning is a valid use case, but it is not intended to be used like this, that why the baseclass throws

We can discuss in person if you want